### ISSUE TITLE
[EXPLICIT] Keyword return (needless_return and implicit_return)

### ISSUE THREAD

**Issue by xavetar** [rust-lang/rust-clippy#13761]
### Summary

In an attempt to prohibit any implicit definitions and language constructs that do not comply with the properties of explicitness, such as return - which for some reason is usually omitted, clippy does not generate either a warning or an error.

How forbid any IMPLICIT representations of these language constructs? Why does Rust allow the use of these IMPLICIT constructs? Where did the mania for reducing everything possible come from? Why is there a keyword "allow" but not a keyword "require"?

### Reproducer

lib.rs:

```rust
#![allow(clippy::needless_return)]
#![forbid(clippy::implicit_return)]
```

### Version

```text
Current Stable: 1.83.0
```

**Comment by samueltardieu** [comment]
> Where did the mania for reducing everything possible come from?

You will find this in Lisp, in the 1960s, and in every functional language since. This may even predate Lisp.

> Why is there a keyword "allow" but not a keyword "require"?

Clippy uses [lint levels from the Rust compiler](https://doc.rust-lang.org/rustc/lints/levels.html). Your question about lacking `require` would be better addressed there.

**Comment by y21** [comment]
Blocks are expressions and its tail expression is what the block evaluates to. If you think of a function body as being a block, then it naturally follows that the tail expression of the function body block is what the function will return.

That said, some people have different opinions and dislike it regardless of whether it makes sense from a language design point of view, and that's fine and is also why we have some of the restriction lints. I'm a bit confused about what this is requesting though. The reproducer has no code other than enabling `implicit_return`. Do you have a case where that lint doesn't emit a warning and you want to report a false negative?

**Comment by xavetar** [comment]
> Blocks are expressions and its tail expression is what the block evaluates to.

Look, from this position, it can be considered acceptable. You can talk all you want about absolute mathematical representation of S-forms, but the user interacts with the language, not with the underlying representation. The result of this expression is machine code, which in its turn contains a return instruction. We write code in a language, the language obeys the rules of linearity, of sequential presentation. When we talk about machine representation, mathematical representation and language representation, we are talking about different entities.

> That said, some people have different opinions and dislike it regardless of whether it makes sense from a language design point of view, and that's fine and is also why we have some of the restriction lints.

I have no opinions, such behavior of the language causes complete irritation (to put it mildly). The language tries to sit on several chairs, simultaneously believing that it is independent of the machine representation and at the same time being completely dependent on concepts created earlier by others, the language cannot separate these representations - it is not in its capabilities.

> I'm a bit confused about what this is requesting though. The reproducer has no code other than enabling `implicit_return`. Do you have a case where that lint doesn't emit a warning and you want to report a false negative?

The question posed by this issue is to make it so that IMPLICIT constructs (return) CAN BE disallowed/restricted, if someone wants to write IMPLICIT code, that's their problem:

```
#![deny(clippy::implicit_return)] | #![forbid(clippy::implicit_return)]
```

**Comment by xavetar** [comment]
> Closing since the `implicit_return` lint already exists.

What exists? Maybe you test with deny and forbid? Or do you have the same logic problems as clippy?

```
#![deny(clippy::implicit_return)] | #![forbid(clippy::implicit_return)]
```
